src/Quartz.Extensions.Hosting/QuartzHostedService.cs

```diff
@@ -1,44 +1,94 @@
 using System.Threading;
 using System.Threading.Tasks;
-
 using Microsoft.Extensions.Hosting;
 using Microsoft.Extensions.Options;
 
 namespace Quartz
 {
     internal class QuartzHostedService : IHostedService
     {
+        private readonly IApplicationLifetime applicationLifetime;
         private readonly ISchedulerFactory schedulerFactory;
         private readonly IOptions<QuartzHostedServiceOptions> options;
         private IScheduler scheduler = null!;
+        private bool schedulerWasStarted;
 
         public QuartzHostedService(
+            IApplicationLifetime applicationLifetime,
             ISchedulerFactory schedulerFactory,
             IOptions<QuartzHostedServiceOptions> options)
         {
+            this.applicationLifetime = applicationLifetime;
             this.schedulerFactory = schedulerFactory;
             this.options = options;
         }
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
+            // Require successful initialization for application startup to succeed
             scheduler = await schedulerFactory.GetScheduler(cancellationToken);
 
+            if (options.Value.AwaitApplicationStarted) // Sensible mode: proceed with startup, and have jobs start after application startup
+            {
+                // Follow the pattern from BackgroundService.StartAsync: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs
+
+                var deferredSchedulerStartTask = AwaitStartupCompletionAndStartSchedulerAsync(cancellationToken);
+
+                // If the task completed synchronously, await it in order to bubble potential cancellation/failure to the caller
+                // Otherwise, return, allowing application startup to complete
+                if (deferredSchedulerStartTask.IsCompleted)
+                {
+                    await deferredSchedulerStartTask.ConfigureAwait(false);
+                }
```

Comment:
I don't think this pattern is quite correct 🤔 If `!deferredSchedulerStartTask.IsCompleted`, I think you need to store that task in a field and await it in `StopAsync`, similar to [the way `BackgroundService` works](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs)?

Reply:
I think you're right. I'll make a change.

(For the sake of completeness, note that this is quite niche. It is only for one scenario: when the application is stopped _while our deferred startup task has not yet completed_. It achieves only one thing: whilst we are cancelling, the await in `StopAsync` helps shutdown wait until our cancellation completes. For `BackgroundService` it is much more relevant, as its execute task is expected to keep running during the application lifetime.)